src/lib/PortalGeneric.tsx

```diff
@@ -0,0 +1,132 @@
+import * as PropTypes from 'prop-types'
+import * as React from 'react'
+import * as _ from 'lodash'
+
+import { eventStack, doesNodeContainClick, AutoControlledComponent } from '.'
+import { Extendable } from 'utils'
+import Ref from '../components/Ref'
+import PortalInner from '../components/Portal/PortalInner'
+
+type ReactMouseEvent = React.MouseEvent<HTMLElement>
+
+export interface IPortalGenericProps {
+  onMount?: (props: IPortalGenericProps) => void
+  onUnmount?: (props: IPortalGenericProps) => void
+  open?: boolean
+}
+
+export interface IPortalGenericState {
+  open?: boolean
+}
+
+/**
+ * A component that allows you to render children outside their parent.
+ */
+export abstract class PortalGeneric<
+  P extends IPortalGenericProps,
+  S extends IPortalGenericState
+> extends AutoControlledComponent<Extendable<P>, S> {
+  protected portalNode: HTMLElement
+  protected triggerNode: HTMLElement
+
+  public static propTypes = {
+    /** Controls whether or not the portal is displayed. */
+    open: PropTypes.bool,
+
+    /** Initial open value */
+    defaultOpen: PropTypes.bool,
+
+    /**
+     * Called when the portal is mounted on the DOM.
+     *
+     * @param {object} data - All props.
+     */
+    onMount: PropTypes.func,
+
+    /**
+     * Called when the portal is unmounted from the DOM.
+     *
+     * @param {object} data - All props.
+     */
+    onUnmount: PropTypes.func,
+  }
+
+  static autoControlledProps = ['open']
+
+  getInitialAutoControlledState() {
```

Comment:
is this meant to be public?

Reply:
right, it can be private. It is called from AutoControlledComponent
```js
 const state = _.invoke(this, 'getInitialAutoControlledState', this.props) || {}
```